spec/an_update_metadata_source_spec.rb

```diff
@@ -0,0 +1,142 @@
+require 'spec_helper'
+
+# See also `gp_update_metadata_source_spec.rb` for the iOS counterpart and an
+# annotated spec.
+#
+# The fact that we have an Android and iOS differentiation is not ideal and
+# something we hope to address in an upcoming refactor/rewrite.
+describe Fastlane::Actions::AnUpdateMetadataSourceAction do
+  it 'updates any block in a given .po file with the values from the given sources' do
+    Dir.mktmpdir do |dir|
+      output_path = File.join(dir, 'output.po')
+      dummy_text = <<~PO
+        msgctxt "release_note_0122"
+        msgid "previous version notes required to have current one added"
+        msgstr ""
+        msgctxt "release_note_0121"
+        msgid "this older release notes block should be removed"
+        msgstr ""
+        msgctxt "key1"
+        msgid "this value should change"
+        msgstr ""
+        msgctxt "key2"
+        msgid "this value should change"
+        msgstr ""
+      PO
+      File.write(output_path, dummy_text)
+
+      release_notes_path = File.join(dir, 'release_notes.txt')
+      File.write(release_notes_path, "- release notes\n- more release notes")
+      file_1_path = File.join(dir, '1.txt')
+      File.write(file_1_path, 'value 1')
+      file_2_path = File.join(dir, '2.txt')
+      File.write(file_2_path, 'value 2')
+
+      described_class.run(
```

Comment:
I know that calling `run` directly is what we've always been doing in our specs that test a fastlane action directly (as opposed to testing just a helper), so it's fine;

But I recently got caught up in a case where my test crashed when calling `run` because I failed to pass a parameter that was supposed to be optional. The issue is that when you call `run` directly, this does not go thru the path in fastlane's code which evaluates the `ConfigItem` and fill `params` with the default values of missing parameters and do the param validation checks etc.

So instead I've recently started to do it differently in my new specs testing fastlane actions, inspired by how Fastlane's own specs do it in the fastlane repo. See https://github.com/wordpress-mobile/release-toolkit/pull/329#discussion_r792028262 for a peek at the technique.

Given those specific tests might (?) disappear soon™ alongside the corresponding actions, and be replaced by a new action and tests replacing them, and also given the test is not really about testing optional parameters or ConfigItem parsing/validation, I think it's fine to keep those like that (i.e. calling `run` directly) for those.
But just wanted to let you know about the alternate way (and the one used by fastlane itself in their test suite) for information and in case you saw a benefit in adopting that new technique in the future 🤷‍♂️

Reply:
I think I'll have a go at using the approach you suggest, even if just to get comfortable with it and set it up for future work.